plugins/hls-explicit-fixity-plugin/src/Ide/Plugin/ExplicitFixity.hs

```diff
@@ -0,0 +1,160 @@
+import qualified Data.Map                             as M
+import           Data.Maybe
+import           Data.Monoid
+import qualified Data.Text                            as T
+import           Development.IDE                      hiding (pluginHandlers,
+                                                       pluginRules)
+import           Development.IDE.Core.PositionMapping (idDelta)
+import           Development.IDE.Core.Shake           (addPersistentRule)
+import qualified Development.IDE.Core.Shake           as Shake
+import           Development.IDE.GHC.Compat
+import           Development.IDE.GHC.Compat.Util      (FastString)
+import qualified Development.IDE.GHC.Compat.Util      as Util
+import           GHC.Generics                         (Generic)
+import           Ide.PluginUtils                      (getNormalizedFilePath,
+                                                       handleMaybeM,
+                                                       pluginResponse)
+import           Ide.Types                            hiding (pluginId)
+import           Language.LSP.Types
+
+pluginId :: PluginId
+pluginId = "explicitFixity"
+
+descriptor :: Recorder (WithPriority Log) -> PluginDescriptor IdeState
+descriptor recorder = (defaultPluginDescriptor pluginId)
+    { pluginRules = fixityRule recorder
+    , pluginHandlers = mkPluginHandler STextDocumentHover hover
+    }
+
+hover :: PluginMethodHandler IdeState TextDocumentHover
+hover state plId (HoverParams (TextDocumentIdentifier uri) pos _) = pluginResponse $ do
+    nfp <- getNormalizedFilePath plId uri
+    fixityTrees <- handleMaybeM "ExplicitFixity: Unable to get fixity"
+        $ liftIO
+        $ runAction "ExplicitFixity.GetFixity" state
+        $ use GetFixity nfp
+    -- We don't have much fixities on one position, so `nubOn` is acceptable.
+    pure $ toHover $ nubOn snd $ concat $ findInTree fixityTrees pos fNodeFixty
+    where
+        toHover [] = Nothing
+        toHover fixities =
+            let contents = T.intercalate "\n\n" $ fixityText <$> fixities
+                contents' = "\n" <> sectionSeparator <> contents
+            in  Just $ Hover (HoverContents $ unmarkedUpContent contents') Nothing
+
+        fixityText (name, Fixity _ precedence direction) =
+            printOutputable direction <> " " <> printOutputable precedence <> " `" <> printOutputable name <> "`"
+
+-- | Transferred from ghc `selectSmallestContaining`
+selectSmallestContainingForFixityTree :: Span -> FixityTree -> Maybe FixityTree
+selectSmallestContainingForFixityTree sp node
+    | sp `containsSpan` fNodeSpan node = Just node
+    | fNodeSpan node `containsSpan` sp = getFirst $ mconcat
+        [ foldMap (First . selectSmallestContainingForFixityTree sp) $ fNodeChildren node
+        , First (Just node)
+        ]
+    | otherwise = Nothing
+
+-- | Transferred from ghcide `pointCommand`
+findInTree :: FixityTrees -> Position -> (FixityTree -> a) -> [a]
```

Comment:
This function has a fairly specific usage, so why do we need this general type? I mean, by using `FixityTree -> [a]` as an argument, we can remove one indirection of `Maybe` and `catMaybe`.